code/StringTagField.php

```diff
@@ -0,0 +1,283 @@
+	/**
+	 * @var array
+	 */
+	public static $allowed_actions = array(
+		'suggest',
+	);
+
+	/**
+	 * @var bool
+	 */
+	protected $shouldLazyLoad = false;
+
+	/**
+	 * @var int
+	 */
+	protected $lazyLoadItemLimit = 10;
+
+	/**
+	 * @var null|DataObject
+	 */
+	protected $record;
+
+	/**
+	 * @param string $name
+	 * @param string $title
+	 * @param array|SS_List $source
+	 * @param array|SS_List $value
+	 */
+	public function __construct($name, $title = '', $source = array(), $value = array()) {
+		parent::__construct($name, $title, $source, $value);
+	}
+
+	/**
+	 * @return bool
+	 */
+	public function getShouldLazyLoad() {
+		return $this->shouldLazyLoad;
+	}
+
+	/**
+	 * @param bool $shouldLazyLoad
+	 *
+	 * @return static
+	 */
+	public function setShouldLazyLoad($shouldLazyLoad) {
+		$this->shouldLazyLoad = $shouldLazyLoad;
+
+		return $this;
+	}
+
+	/**
+	 * @return int
+	 */
+	public function getLazyLoadItemLimit() {
+		return $this->lazyLoadItemLimit;
+	}
+
+	/**
+	 * @param int $lazyLoadItemLimit
+	 *
+	 * @return static
+	 */
+	public function setLazyLoadItemLimit($lazyLoadItemLimit) {
+		$this->lazyLoadItemLimit = $lazyLoadItemLimit;
+
+		return $this;
+	}
+
+	/**
+	 * @return null|DataObject
+	 */
+	public function getRecord() {
+		if($this->record) {
+			return $this->record;
+		}
+
+		if($form = $this->getForm()) {
+			return $form->getRecord();
+		}
+
+		return null;
+	}
+
+	/**
+	 * @param DataObject $record
+	 *
+	 * @return $this
+	 */
+	public function setRecord(DataObject $record) {
+		$this->record = $record;
+
+		return $this;
+	}
+
+	/**
+	 * {@inheritdoc}
+	 */
+	public function Field($properties = array()) {
+		Requirements::css(TAG_FIELD_DIR . '/css/select2.min.css');
+		Requirements::css(TAG_FIELD_DIR . '/css/TagField.css');
+
+		Requirements::javascript(THIRDPARTY_DIR . '/jquery/jquery.js');
+		Requirements::javascript(THIRDPARTY_DIR . '/jquery-entwine/dist/jquery.entwine-dist.js');
+		Requirements::javascript(TAG_FIELD_DIR . '/js/select2.js');
+		Requirements::javascript(TAG_FIELD_DIR . '/js/TagField.js');
+
+		$this->addExtraClass('silverstripe-tag-field');
+
+		$this->setAttribute('multiple', 'multiple');
+
+		if($this->getShouldLazyLoad()) {
+			$this->setAttribute('data-silverstripe-tag-field-suggest-url', $this->getSuggestURL());
+		} else {
+			$properties = array_merge($properties, array(
+				'Options' => $this->getOptions()
+			));
+		}
+
+		return $this
+			->customise($properties)
+			->renderWith(array("templates/StringTagField"));
+	}
+
+	/**
+	 * @return string
+	 */
+	protected function getSuggestURL() {
+		return Controller::join_links($this->Link(), 'suggest');
+	}
+
+	/**
+	 * @return ArrayList
+	 */
+	protected function getOptions() {
+		$options = ArrayList::create();
+
+		$source = $this->getSource();
+
+		if($source instanceof Iterator) {
+			$source = iterator_to_array($source);
+		}
+
+		$values = $this->Value();
+
+		foreach($source as $key => $value) {
```

Comment:
The README example defines the `$source` as `array('one', 'two')`. You're looping over key/val here, which means the keys will be numeric. Yet in `suggest()` you're setting tags with `'id' => $tag`, so an associative array. Is the example wrong, or am I missing something here?